samples/routing/custom/Core_6/Shared/AutomaticRouting/RoutingInfoSubscriber.cs

```diff
@@ -0,0 +1,331 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using NServiceBus;
+using NServiceBus.Features;
+using NServiceBus.Logging;
+using NServiceBus.Routing;
+using NServiceBus.Routing.MessageDrivenSubscriptions;
+
+class RoutingInfoSubscriber : FeatureStartupTask
+{
+    static ILog log = LogManager.GetLogger<RoutingInfoSubscriber>();
+
+    UnicastRoutingTable routingTable;
+    EndpointInstances endpointInstances;
+    IReadOnlyCollection<Type> messageTypesHandledByThisEndpoint;
+    Publishers publishers;
+    TimeSpan sweepPeriod;
+    TimeSpan heartbeatTimeout;
+    Dictionary<Type, HashSet<string>> endpointMap = new Dictionary<Type, HashSet<string>>();
+    Dictionary<string, HashSet<EndpointInstance>> instanceMap = new Dictionary<string, HashSet<EndpointInstance>>();
+    Dictionary<EndpointInstance, EndpointInstanceInfo> instanceInformation = new Dictionary<EndpointInstance, EndpointInstanceInfo>();
+    Dictionary<Type, string> publisherMap = new Dictionary<Type, string>();
+    Timer sweepTimer;
+    IMessageSession messageSession;
+
+    public RoutingInfoSubscriber(UnicastRoutingTable routingTable, EndpointInstances endpointInstances, IReadOnlyCollection<Type> messageTypesHandledByThisEndpoint, Publishers publishers, TimeSpan sweepPeriod, TimeSpan heartbeatTimeout)
+    {
+        this.routingTable = routingTable;
+        this.endpointInstances = endpointInstances;
+        this.messageTypesHandledByThisEndpoint = messageTypesHandledByThisEndpoint;
+        this.publishers = publishers;
+        this.sweepPeriod = sweepPeriod;
+        this.heartbeatTimeout = heartbeatTimeout;
+    }
+
+    protected override Task OnStart(IMessageSession context)
+    {
+        messageSession = context;
+
+        #region AddDynamic
+        routingTable.AddDynamic((list, bag) => FindEndpoint(list));
```

Comment:
Maybe not the best place to comment on that but I think "AddDynamic" is not the best name for this method, at first look I was very confused. I think it'd be better as simply 'Add' (== another overload of the method that just adds an endpoint), even better if both would be called 'AddEdnpoint'.

Personal opinion.